File: solved/isPalindrome.ts
import { ListNode } from '../lib/ListNode'

// https://leetcode.com/explore/learn/card/linked-list/219/classic-problems/1209/
type MaybeNode = ListNode | null
const isPalindrome = (head: MaybeNode): boolean => {
  let cur = head
  if (!cur) return true
  const st = [] as number[]

  while (cur) {
    st.push(cur.val)
    cur = cur.next
  }

  let i = 0
  let j = st.length - 1
  while (i < j) if (st[i++] !== st[j--]) return false

  return true
}

export { isPalindrome }

File: solved/go/IsPalindrome/IsPalindrome.go
// Package main has solved algorithms
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/93/linked-list/772/
package main

// import log "github.com/sirupsen/logrus"

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	st := make([]int, 0)
	slow := head
	fast := head

	for fast != nil && fast.Next != nil {
		st = append(st, slow.Val)
		slow = slow.Next
		fast = fast.Next.Next
	}

	if fast != nil {
		slow = slow.Next
	}

	for i := len(st) - 1; i >= 0; i-- {
		if st[i] != slow.Val {
			return false
		}
		slow = slow.Next
	}

	return true
}

func main() {
	// log.Info(isPalindrome())
}
